std/protocols/fingerprint.asm

```diff
@@ -1,24 +1,52 @@
+use std::array;
 use std::array::len;
-use std::utils::fold;
 use std::math::fp2::Fp2;
 use std::math::fp2::add_ext;
 use std::math::fp2::mul_ext;
 use std::math::fp2::pow_ext;
 use std::math::fp2::from_base;
+use std::math::fp2::eval_ext;
+use std::check::assert;
 
 /// Maps [x_1, x_2, ..., x_n] to its Read-Solomon fingerprint, using a challenge alpha: $\sum_{i=1}^n alpha**{(n - i)} * x_i$
-let<T: Add + Mul + FromLiteral> fingerprint: T[], Fp2<T> -> Fp2<T> = |expr_array, alpha| {
-    let n = len(expr_array);
-    fold(
-        n,
-        |i| mul_ext(pow_ext(alpha, n - i - 1), from_base(expr_array[i])),
-        from_base(0),
-        |sum_acc, el| add_ext(sum_acc, el)
-    )
+/// Note that while this function *would* work for expressions, it would generate exponentially large results.
```

Comment:
Is it correct that this function is only used in the context of prover functions? If so, it would make sense to indicate that here

Reply:
It is indicated with the `query` keyword, right? Do you think it needs an extra comment?

Reply:
I simplified the comment a bit!

The function used to be generic over any `T: Add + Mul + FromLiteral`, which includes expression. I changed that, not because it wouldn't work anymore, but so that using this function to generate an expression would cause a compile error. The comment made more sense when it was still generic.